File: sources/c/14_file_read.c
// task 14 file_read — expected output: 484442112
// build: gcc -O2 -pthread -o prog 14_file_read.c    run: ./prog
// alternates: clang -O2 -pthread -o prog 14_file_read.c | cl /O2 /Fe:prog 14_file_read.c | tcc -o prog 14_file_read.c

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

#define CHUNK 1048576   /* 1 MiB */

int main(void) {
    FILE *f = fopen("data.bin", "rb");
    if (f == NULL) {
        return 1;
    }

    unsigned char *buf = (unsigned char *)malloc(CHUNK);
    uint64_t total = 0;
    size_t got;

    while ((got = fread(buf, 1, CHUNK, f)) > 0) {
        for (size_t i = 0; i < got; i++) {
            total += (uint64_t)buf[i];
        }
    }

    fclose(f);
    free(buf);

    printf("%llu\n", (unsigned long long)(total % 4294967296ULL));
    return 0;
}

File: sources/php/14_file_read.php
<?php
// task 14 file_read — expected output: 484442112
// build: none (interpreted)    run: php 14_file_read.php (zend) | php -d opcache.enable_cli=1 -d opcache.jit_buffer_size=64M 14_file_read.php (zend + jit)
// Reads data.bin (104857600 bytes: the bytes 0..255 repeating) from the working directory in
// 1 MiB chunks, never one byte per syscall, and sums every byte.

$handle = fopen('data.bin', 'rb');

$total = 0;

while (!feof($handle)) {
    $chunk = fread($handle, 1048576);
    if ($chunk === false || $chunk === '') {
        break;
    }
    $len = strlen($chunk);
    for ($i = 0; $i < $len; $i++) {
        $total += ord($chunk[$i]);
    }
}

fclose($handle);

echo $total % 4294967296, "\n";
